helper/ipsec.c

```diff
@@ -0,0 +1,157 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright(C) 2020 Marvell.
+ */
+
+#include <odp/helper/ipsec.h>
+#include <odp/helper/odph_debug.h>
+
+int odph_ipsec_capability_check(odp_bool_t ah,
+				odp_cipher_alg_t cipher_alg,
+				uint32_t cipher_key_len,
+				odp_auth_alg_t auth_alg,
+				uint32_t auth_key_len)
+{
+	odp_ipsec_capability_t capa;
+	odp_ipsec_cipher_capability_t cipher_capa[MAX_ALG_CAPA];
+	odp_ipsec_auth_capability_t   auth_capa[MAX_ALG_CAPA];
+	int i, num;
+	odp_bool_t found;
+
+	if (odp_ipsec_capability(&capa) < 0)
+		return -1;
+
+	if (ah && (ODP_SUPPORT_NO == capa.proto_ah))
+		return -1;
+
+	/* Cipher algorithms */
+	switch (cipher_alg) {
+	case ODP_CIPHER_ALG_NULL:
+		if (!capa.ciphers.bit.null)
+			return -1;
+		break;
+	case ODP_CIPHER_ALG_DES:
+		if (!capa.ciphers.bit.des)
+			return -1;
+		break;
+	case ODP_CIPHER_ALG_3DES_CBC:
+		if (!capa.ciphers.bit.trides_cbc)
+			return -1;
+		break;
+	case ODP_CIPHER_ALG_AES_CBC:
+		if (!capa.ciphers.bit.aes_cbc)
+			return -1;
+		break;
+	case ODP_CIPHER_ALG_AES_CTR:
+		if (!capa.ciphers.bit.aes_ctr)
+			return -1;
+		break;
+	case ODP_CIPHER_ALG_AES_GCM:
+		if (!capa.ciphers.bit.aes_gcm)
+			return -1;
+		break;
+	case ODP_CIPHER_ALG_AES_CCM:
+		if (!capa.ciphers.bit.aes_ccm)
+			return -1;
+		break;
+	case ODP_CIPHER_ALG_CHACHA20_POLY1305:
+		if (!capa.ciphers.bit.chacha20_poly1305)
+			return -1;
+		break;
+	default:
+		ODPH_ERR("Unsupported cipher algorithm\n");
+		return -1;
+	}
+
+	/* Authentication algorithms */
+	switch (auth_alg) {
+	case ODP_AUTH_ALG_NULL:
+		if (!capa.auths.bit.null)
+			return -1;
+		break;
+	case ODP_AUTH_ALG_MD5_HMAC:
+		if (!capa.auths.bit.md5_hmac)
+			return -1;
+		break;
+	case ODP_AUTH_ALG_SHA1_HMAC:
+		if (!capa.auths.bit.sha1_hmac)
+			return -1;
+		break;
+	case ODP_AUTH_ALG_SHA256_HMAC:
+		if (!capa.auths.bit.sha256_hmac)
+			return -1;
+		break;
+	case ODP_AUTH_ALG_SHA384_HMAC:
+		if (!capa.auths.bit.sha384_hmac)
+			return -1;
+		break;
+	case ODP_AUTH_ALG_SHA512_HMAC:
+		if (!capa.auths.bit.sha512_hmac)
+			return -1;
+		break;
+	case ODP_AUTH_ALG_AES_XCBC_MAC:
+		if (!capa.auths.bit.aes_xcbc_mac)
+			return -1;
+		break;
+	case ODP_AUTH_ALG_AES_GCM:
+		if (!capa.auths.bit.aes_gcm)
+			return -1;
+		break;
+	case ODP_AUTH_ALG_AES_GMAC:
+		if (!capa.auths.bit.aes_gmac)
+			return -1;
+		break;
+	case ODP_AUTH_ALG_AES_CCM:
+		if (!capa.auths.bit.aes_ccm)
+			return -1;
+		break;
+	case ODP_AUTH_ALG_CHACHA20_POLY1305:
+		if (!capa.auths.bit.chacha20_poly1305)
+			return -1;
+		break;
+	default:
+		ODPH_ERR("Unsupported authentication algorithm\n");
+		return -1;
+	}
+
+	num = odp_ipsec_cipher_capability(cipher_alg, cipher_capa, MAX_ALG_CAPA);
+	if (num <= 0) {
+		ODPH_ERR("Could not get cipher capabilities\n");
+		return -1;
+	}
+
+	/* Search for the test case */
+	found = false;
+	for (i = 0; i < num; i++) {
+		if (cipher_capa[i].key_len == cipher_key_len) {
```

Comment:
This does out-of-bounds array access if the number of cipher capabilities exceed the size of the array. Note that odp_ipsec_cipher_capability() returns the total number of capabilities there are, even if it is larger than the size of the supplied array.

Reply:
Ack. Updated the function to use dynamic sized array.